File: src/main/java/com/tabnine/inline/render/experimental/ExperimentalTabnineInlay.java
package com.tabnine.inline.render.experimental;

import com.intellij.openapi.Disposable;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.util.Disposer;
import com.tabnine.general.Utils;
import com.tabnine.inline.render.GenericInlayWrapper;
import com.tabnine.inline.render.TabnineInlay;
import com.tabnine.prediction.TabNineCompletion;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import java.awt.*;
import java.util.List;
import java.util.stream.Collectors;

public class ExperimentalTabnineInlay implements TabnineInlay {
    private GenericInlayWrapper inlineBeforeSuffix;
    private GenericInlayWrapper inlineAfterSuffix;
    private GenericInlayWrapper block;

    @Nullable
    public Integer getOffset() {
        if (inlineBeforeSuffix != null) {
            return inlineBeforeSuffix.inner().getOffset();
        }
        if (block != null) {
            return block.inner().getOffset();
        }

        return null;
    }

    @Nullable
    public Rectangle getBounds() {
        if (inlineBeforeSuffix != null) {
            return inlineBeforeSuffix.inner().getBounds();
        }
        if (block != null) {
            return block.inner().getBounds();
        }

        return null;
    }

    public boolean isEmpty() {
        return this.inlineBeforeSuffix == null && this.inlineAfterSuffix == null && this.block == null;
    }

    public void register(Disposable parent) {
        if (inlineBeforeSuffix != null) {
            Disposer.register(parent, inlineBeforeSuffix.inner());
        }
        if (inlineAfterSuffix != null) {
            Disposer.register(parent, inlineAfterSuffix.inner());
        }
        if (block != null) {
            Disposer.register(parent, block.inner());
        }
    }

    public void clear() {
        if (inlineBeforeSuffix != null) {
            Disposer.dispose(inlineBeforeSuffix.inner());
            inlineBeforeSuffix = null;
        }

        if (inlineAfterSuffix != null) {
            Disposer.dispose(inlineAfterSuffix.inner());
            inlineAfterSuffix = null;
        }

        if (block != null) {
            Disposer.dispose(block.inner());
            block = null;
        }
    }

    public void render(Editor editor, @NotNull String suffix, TabNineCompletion completion, int offset) {
        List<String> lines = Utils.asLines(suffix);
        String firstLine = lines.get(0);
        List<String> otherLines = lines.stream().skip(1).collect(Collectors.toList());

        if (!firstLine.isEmpty()) {
            int endIndex = firstLine.indexOf(completion.oldSuffix);
            if (!completion.oldSuffix.isEmpty() && endIndex >= 0) {
                String before = firstLine.substring(0, endIndex);
                InlineElementRenderer beforeInline =
                        new InlineElementRenderer(editor, before, completion.deprecated);
                this.inlineBeforeSuffix = new GenericInlayWrapper(editor
                        .getInlayModel()
                        .addInlineElement(offset, true, beforeInline));

                int startOfAfter = endIndex + completion.oldSuffix.length();
                if (startOfAfter < firstLine.length()) {
                    String after = firstLine.substring(startOfAfter);

                    InlineElementRenderer afterInline =
                            new InlineElementRenderer(editor, after, completion.deprecated);
                    this.inlineAfterSuffix = new GenericInlayWrapper(editor
                            .getInlayModel()
                            .addInlineElement(offset + before.length(), true, afterInline));
                }
            } else {
            InlineElementRenderer inlineElementRenderer =
                    new InlineElementRenderer(editor, firstLine, completion.deprecated);
            this.inlineBeforeSuffix = new GenericInlayWrapper(editor
                    .getInlayModel()
                    .addInlineElement(offset, true, inlineElementRenderer));
            }
        }
        if (otherLines.size() > 0) {
            BlockElementRenderer blockElementRenderer =
                    new BlockElementRenderer(editor, otherLines, completion.deprecated);
            this.block = new GenericInlayWrapper(editor
                    .getInlayModel()
                    .addBlockElement(
                            offset,
                            true,
                            false,
                            1,
                            blockElementRenderer
                    ));
        }
    }
}

File: src/main/java/com/tabnine/inline/render/experimental/ExperimentalTabnineInlay.kt
package com.tabnine.inline.render.experimental

import com.intellij.openapi.Disposable
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.editor.Inlay
import com.intellij.openapi.util.Disposer
import com.tabnine.general.Utils
import com.tabnine.inline.render.TabnineInlay
import com.tabnine.prediction.TabNineCompletion
import java.awt.Rectangle
import java.util.stream.Collectors

class ExperimentalTabnineInlay : TabnineInlay {
    private var beforeSuffixInlay: Inlay<*>? = null
    private var afterSuffixInlay: Inlay<*>? = null
    private var blockInlay: Inlay<*>? = null

    override val offset: Int?
        get() = beforeSuffixInlay?.offset ?: afterSuffixInlay?.offset ?: blockInlay?.offset

    override fun getBounds(): Rectangle? {
        val result = beforeSuffixInlay?.bounds?.let { Rectangle(it) }

        result?.bounds?.let {
            afterSuffixInlay?.bounds?.let { after -> result.add(after) }
            blockInlay?.bounds?.let { blockBounds -> result.add(blockBounds) }
        }

        return result
    }

    override val isEmpty: Boolean
        get() = beforeSuffixInlay == null && afterSuffixInlay == null && blockInlay == null

    override fun register(parent: Disposable) {
        beforeSuffixInlay?.let {
            Disposer.register(parent, it)
        }
        afterSuffixInlay?.let {
            Disposer.register(parent, it)
        }
        blockInlay?.let {
            Disposer.register(parent, it)
        }
    }

    override fun clear() {
        beforeSuffixInlay?.let {
            Disposer.dispose(it)
            beforeSuffixInlay = null
        }
        afterSuffixInlay?.let {
            Disposer.dispose(it)
            afterSuffixInlay = null
        }
        blockInlay?.let {
            Disposer.dispose(it)
            blockInlay = null
        }
    }

    override fun render(editor: Editor, suffix: String, completion: TabNineCompletion, offset: Int) {
        val lines = Utils.asLines(suffix)
        val firstLine = lines[0]
        val otherLines = lines.stream().skip(1).collect(Collectors.toList())
        if (firstLine.isNotEmpty()) {
            val endIndex = firstLine.indexOf(completion.oldSuffix)
            if (completion.oldSuffix.isNotEmpty() && endIndex > 0) {
                val beforeSuffix = firstLine.substring(0, endIndex)
                beforeSuffixInlay = renderInline(editor, beforeSuffix, completion, offset)

                val afterSuffixIndex = endIndex + completion.oldSuffix.length
                val after = if (afterSuffixIndex < firstLine.length) firstLine.substring(afterSuffixIndex) else null
                after?.let {
                    afterSuffixInlay = renderInline(editor, it, completion, offset + beforeSuffix.length)
                }
            } else {
                beforeSuffixInlay = renderInline(editor, firstLine, completion, offset)
            }
        }
        if (otherLines.size > 0) {
            blockInlay = renderBlock(editor, otherLines, completion, offset)
        }
    }

    private fun renderBlock(
        editor: Editor,
        otherLines: MutableList<String>,
        completion: TabNineCompletion,
        offset: Int
    ): Inlay<BlockElementRenderer>? {
        val blockElementRenderer = BlockElementRenderer(editor, otherLines, completion.deprecated)
        return editor
            .inlayModel
            .addBlockElement(
                offset,
                true,
                false,
                1,
                blockElementRenderer
            )
    }

    private fun renderInline(
        editor: Editor,
        before: String,
        completion: TabNineCompletion,
        offset: Int
    ): Inlay<InlineElementRenderer>? {
        val inline = InlineElementRenderer(editor, before, completion.deprecated)
        return editor
            .inlayModel
            .addInlineElement(offset, true, inline)
    }
}
